code/controllers/subsystems/documents.dm

```diff
@@ -0,0 +1,242 @@
+#define SSDOCS_MEDIUM_PAPER "paper"
+#define SSDOCS_MEDIUM_FILE "file"
+
+var/datum/controller/subsystem/docs/SSdocs
+
+/datum/controller/subsystem/docs
+	name = "Documents"
+	wait = 30 SECONDS
+	flags = SS_NO_FIRE
+	init_order = SS_INIT_MISC_FIRST
+
+	var/total_docs = 0
+	var/list/docs = list()
+	var/list/docs_by_tags = list()
+	var/list/total_by_tags = list()
+
+/datum/controller/subsystem/docs/Recover()
+	src.docs = SSdocs.docs
+	src.docs_by_tags = SSdocs.docs_by_tags
+
+/datum/controller/subsystem/docs/Initialize(timeofday)
+	//Load in the docs config
+	if(config.docs_load_docs_from == "sql")
+		log_debug("SSdocs: Attempting to Load from SQL")
+		load_from_sql()
+	else if(config.docs_load_docs_from == "json")
+		log_debug("SSdocs: Attempting to Load from JSON")
+		load_from_json()
+	else
+		log_game("SSdocs: invalid load option specified in config")
+
+	..()
+
+/datum/controller/subsystem/docs/New()
+	NEW_SS_GLOBAL(SSdocs)
+
+/*
+	Fetching Data
+*/
+//A rewritten version of pickweight.
+/datum/controller/subsystem/docs/proc/pick_document()
+	var/subtotal = rand() * src.total_docs
+	var/doc
+	for (doc in docs)
+		subtotal -= docs[doc].chance
+		if (subtotal <= 0)
+			return docs[doc]
+	return 0
+
+//Pick a document by one tag
+/datum/controller/subsystem/docs/proc/pick_document_by_tag(var/tag)
+	if(!docs_by_tags[tag])
+		return 0
+	var/subtotal = rand() * src.total_by_tags[tag]
+	var/doc
+	for (doc in docs_by_tags[tag])
+		subtotal -= docs_by_tags[tag][doc].chance
+		if (subtotal <= 0)
+			return docs_by_tags[tag][doc]
+	return 0
+
+//Pick a document by any tag from a list of tags. Weighted.
+/datum/controller/subsystem/docs/proc/pick_document_by_any_tag(var/list/tags)
+	if(!istype(tags) || !tags.len)
+		return 0
+	var/total_chance = 0
+	var/tag_sublist = list()
+	for(var/t in tags)
+		if(!total_by_tags[t])
+			return
+		total_chance += total_by_tags[t]
+		tag_sublist += docs_by_tags[t]
+	var/subtotal = total_chance * rand()
+	var/doc
+	for(doc in tag_sublist)
+		subtotal -= tag_sublist[doc].chance
+		if(subtotal <= 0)
+			return tag_sublist[doc]
+	return 0
+
+//Pick a document by multiple tags that it must have.
+/datum/controller/subsystem/docs/proc/pick_document_by_tags(var/list/tags)
+	if(!istype(tags) || !tags.len)
+		return 0
+	var/list/tag_sublist = docs_by_tags[pick(tags)] // the list cannot start off as empty
+	for(var/t in tags)
+		if(!docs_by_tags[t])
+			return 0
+		tag_sublist &= docs_by_tags[t]
+	log_ss("docs", "Tag sublist has length [tag_sublist.len].")
+	var/subtotal = 0
+	var/datum/docs_document/dd
+	for(dd in tag_sublist)
+		if(!istype(dd))
```

Comment:
Unnecessary. `for (var/path/x in y)` already executes an `istype()` check, and simply neglects iteration if the type does not match.